src/trans/gpu/internal/dir_trans_ctl_mod.F90

```diff
@@ -9,214 +10,175 @@
+    REAL(KIND=JPRB)    ,OPTIONAL, INTENT(IN)  :: PGP3B(:,:,:,:)
+    REAL(KIND=JPRB)    ,OPTIONAL, INTENT(IN)  :: PGP2(:,:,:)
+
+    ! Local variables
+
+    INTEGER(KIND=JPIM) :: IPTRGP(KF_GP),IPTRSPUV(NPROMATR),IPTRSPSC(NPROMATR)
+    INTEGER(KIND=JPIM) :: ISHFUV_G(KF_GP),ISHFSC_G(KF_GP)
+    INTEGER(KIND=JPIM) :: IVSETUV(KF_GP),IVSETSC(KF_GP)
+    INTEGER(KIND=JPIM) :: IBLKS,JBLK,ISTUV_G,IENUV_G
+    INTEGER(KIND=JPIM) :: IF_UV_G,IF_UV,ISTUV,IF_SCALARS,IF_SCALARS_G,IF_FS,IF_GP
+    INTEGER(KIND=JPIM) :: JFLD,ISTSC_G,IENSC_G,ISTSC,IENSC,IENUV,IF_GPB
+
+    REAL(KIND=JPRBT), POINTER :: FOUBUF_IN(:), FOUBUF(:)
+    REAL(KIND=JPRBT), POINTER :: PREEL_REAL(:), PREEL_COMPLEX(:)
+
+    REAL(KIND=JPRBT), POINTER :: ZINPS(:), ZINPA(:)
+    REAL(KIND=JPRD), POINTER :: ZINPS0(:), ZINPA0(:)
+
+    TYPE(BUFFERED_ALLOCATOR) :: ALLOCATOR
+    TYPE(TRGTOL_HANDLE) :: HTRGTOL
+    TYPE(FTDIR_HANDLE) :: HFTDIR
+    TYPE(TRLTOM_PACK_HANDLE) :: HTRLTOM_PACK
+    TYPE(TRLTOM_HANDLE) :: HTRLTOM
+    TYPE(TRLTOM_UNPACK_HANDLE) :: HTRLTOM_UNPACK
+    TYPE(LTDIR_HANDLE) :: HLTDIR
+
+    IF(NPROMATR > 0) THEN
+        PRINT *, "ERROR, not implemented right now (NPROMATR > 0)"
+        STOP 4
+    ENDIF
 
-!     Method.
-!     -------
+    ! Prepare everything
+    ALLOCATOR = MAKE_BUFFERED_ALLOCATOR()
+    HTRGTOL = PREPARE_TRGTOL(ALLOCATOR,KF_GP,KF_FS)
+    HFTDIR = PREPARE_FTDIR()
+    HTRLTOM_PACK = PREPARE_TRLTOM_PACK(ALLOCATOR, KF_FS)
+    HTRLTOM = PREPARE_TRLTOM(ALLOCATOR, KF_FS)
+    HTRLTOM_UNPACK = PREPARE_TRLTOM_UNPACK(ALLOCATOR, KF_FS)
+    HLTDIR = PREPARE_LTDIR(ALLOCATOR, KF_FS, KF_UV)
 
-!     Externals.  SHUFFLE     - reshuffle fields for load balancing
-!     ----------  FIELD_SPLIT - split fields in NPROMATR packets
-!                 LTDIR_CTL   - control of Legendre transform
-!                 FTDIR_CTL   - control of Fourier transform
+    CALL INSTANTIATE_ALLOCATOR(ALLOCATOR, REUSE_PTR)
 
-!     Author.
-!     -------
-!        Mats Hamrud *ECMWF*
+    ! from the PGP arrays to PREEL_REAL
+    CALL TRGTOL(ALLOCATOR,HTRGTOL,PREEL_REAL,KF_FS,KF_GP,KF_UV_G,KF_SCALARS_G,&
+     & KVSETUV=KVSETUV,KVSETSC=KVSETSC,&
+     & KVSETSC3A=KVSETSC3A,KVSETSC3B=KVSETSC3B,KVSETSC2=KVSETSC2,&
+     & PGP=PGP,PGPUV=PGPUV,PGP3A=PGP3A,PGP3B=PGP3B,PGP2=PGP2)
 
-!     Modifications.
-!     --------------
-!        Original : 01-01-03
+    IF (KF_FS > 0) THEN
 
-!     ------------------------------------------------------------------
+      ! fourier transform from PREEL_REAL to PREEL_COMPLEX (in-place!)
+      CALL GSTATS(1640,0)
+      CALL FTDIR(HFTDIR,PREEL_REAL,PREEL_COMPLEX,KF_FS)
+      CALL GSTATS(1640,1)
 
-USE PARKIND1  ,ONLY : JPIM     ,JPRB
+      CALL GSTATS(153,0)
 
-USE TPM_GEN         ,ONLY : NPROMATR
-USE TPM_TRANS       ,ONLY : FOUBUF_IN, NF_SC2, NF_SC3A, NF_SC3B
-!USE TPM_DISTR
+      WRITE(NOUT,*) 'ltdir_ctl:TRLTOM_CUDAAWARE'
+      CALL TRLTOM_PACK(ALLOCATOR,HTRLTOM_PACK,PREEL_COMPLEX,FOUBUF_IN,KF_FS)
```

Comment:
note that I moved the pack/unpack routines here on purpose because it makes more sense. The pack/unpack were initially called through TRLTOM/TRMTOG but I think this is not a good decision because these function live in between e.g. FTDIR and TRLTOM without preference for either function.